### ISSUE TITLE
Consider installing tomcat native libraries or provide variant that does.

### ISSUE THREAD

**Issue by dougborg** [docker-library/tomcat#7]
We are currently doing this in our tomcat container ((https://registry.hub.docker.com/u/readytalk/tomcat-native/dockerfile/), and I think it should actually be part of the official one :

```
ENV DEBIAN_FRONTEND noninteractive
ENV LD_LIBRARY_PATH ${LD_LIBRARY_PATH}:/usr/lib/x86_64-linux-gnu/

RUN apt-get update
RUN apt-get install -y libtcnative-1
```

More information on the tomcat native libraries here:
http://tomcat.apache.org/tomcat-8.0-doc/apr.html

From the docs:
>Tomcat can use the Apache Portable Runtime to provide superior scalability, performance, and better integration with native server technologies. The Apache Portable Runtime is a highly portable library that is at the heart of Apache HTTP Server 2.x. APR has many uses, including access to advanced IO functionality (such as sendfile, epoll and OpenSSL), OS level functionality (random number generation, system status, etc), and native process handling (shared memory, NT pipes and Unix sockets).

>These features allows making Tomcat a general purpose webserver, will enable much better integration with other native web technologies, and overall make Java much more viable as a full fledged webserver platform rather than simply a backend focused technology.

I'd be happy to open a pull request if this is something people are interested in.

**Comment by yosifkit** [comment]
This seems fine to me, but I am just the maintainer and not a user.  Any users have an opinion here?  Would a separate tag be better (there are already quite a few of them) or just include it always?

What is the size difference if we add this just before the `RUN gpg --keyserver...` lines?
```Dockerfile
RUN apt-get update \
    && apt-get install -y libtcnative-1 \
    && rm -rf /var/lib/apt/lists/*
```

**Comment by tianon** [comment]
Wouldn't we need to pull libtcnative from upstream instead of from
packages, like how we pull tomcat?

**Comment by alexsantos** [comment]
:+1: 

BTW, from packages I get this INFO:

    INFO: An older version 1.1.32 of the APR based Apache Tomcat Native library is installed, while Tomcat recommends a minimum version of 1.1.33

**Comment by md5** [comment]
Based on @alexsantos's comment, I think the answer to @tianon's question about whether to pull from upstream instead from packages is "Yes". Looks like 1.1.33 is available in Debian Sid, but not in any of the other suites.

**Comment by md5** [comment]
Regarding @yosifkit's comments, I don't think that including `libtcnative` necessitates additional tags. APR support should be built into all versions of the `tomcat` image.

**Comment by yosifkit** [comment]
Cool, anyone up to adding an up-to-date `libtcnative`?

**Comment by md5** [comment]
I don't have time at the moment, but it looks pretty easy. The build instructions are here: https://svn.apache.org/repos/asf/tomcat/native/branches/1.1.x/native/BUILDING

The source package and GPG signature can be downloaded via: https://tomcat.apache.org/download-native.cgi

**Comment by md5** [comment]
I spent a little time tonight getting `libtcnative` to build from source. The result is here: https://github.com/docker-library/tomcat/compare/master...appropriate:libtcnative

I didn't bother to update the `update.sh` script yet because I didn't look into how to determine the right version of `libtcnative` required for a particular version of Tomcat.

You may also notice that I had to install the full JDK in order to build `libtcnative`... It gets removed in the same `RUN` statement, so the resulting layer is about 4 megabytes (it could be smaller by half if I removed the `libtcnative*.a` file that gets installed).

**Comment by md5** [comment]
I just noticed this in the docs:

> The wrapper library sources are located in the Tomcat binary bundle, in the bin/tomcat-native.tar.gz archive. 

That should mean that the approach I took in my demonstration branch of downloading `tomcat-native-$TCNATIVE_VERSION-src.tar.gz` is unnecessary. I don't have time to test right now, but it seems worth investigating.

**Comment by md5** [comment]
OK, I lied. I did at least have time to confirm that the sources are already there:

```console
$ docker run --rm tomcat find /usr/local/tomcat -name '*.tar.gz'
/usr/local/tomcat/bin/commons-daemon-native.tar.gz
/usr/local/tomcat/bin/tomcat-native.tar.gz
```

**Comment by ismaello** [comment]
Just after install tomcat7: 
#No install libnative via apt-get 
RUN apt-get update && \
    apt-get install -yq build-essential openssl libssl-dev  curl


RUN cd /tmp \
    && curl -fSL "http://www.us.apache.org/dist//apr/apr-1.5.2.tar.gz" -O \
    && tar xvzf apr-1.5.2.tar.gz && cd apr-1.5.2 \
    && ./configure --prefix=/usr/local/apr && make && make install


RUN cd /tmp \
    && curl -fSL "http://www.us.apache.org/dist/tomcat/tomcat-connectors/native/1.1.33/source/tomcat-native-1.1.33-src.tar.gz" -O \
    && tar xvzf tomcat-native-1.1.33-src.tar.gz && cd tomcat-native-1.1.33-src/jni/native \
    && ./configure --with-apr=/usr/local/apr && make && make install

ENV LD_LIBRARY_PATH /usr/local/apr/lib:$LD_LIBRARY_PATH

**Comment by md5** [comment]
Thanks @ismaello!

These steps look correct, but I also think installing `libapr1` from the Debian package would be fine. In addition, I believe it's prudent to use the `tomcat-native` source that's bundled with the version of Tomcat used in the image in order to ensure that the correct version is used.

Any chance you'd want to create a PR based on your instructions, but using `apt-get` to install `libapr1` and using the `tomcat-native` tarball at `/usr/local/tomcat/bin/tomcat-native.tar.gz`? You'll need to install and remove `libapr-dev` as well for building the `tomcat-native` source.

**Comment by ahromis** [comment]
@md5 is something like this more of what you were looking for?  It seems be working great on a container I just built.
```
ENV JAVA_JDK_DEBIAN_VERSION 7u85-2.6.1-6+deb8u1

RUN apt-get update \
    && apt-get install -yq build-essential openssl libssl-dev curl libapr1 libapr1-dev \
        openjdk-7-jdk="$JAVA_JDK_DEBIAN_VERSION" \
    && rm -rf /var/lib/apt/lists/* \
    && tar zxf /usr/local/tomcat/bin/tomcat-native.tar.gz -C /tmp && cd /tmp/tomcat-native*-src/jni/native/ \
    && ./configure --with-apr=/usr/bin/apr-config --with-java-home=/usr/lib/jvm/java-7-openjdk-amd64/ \
    && make \
    && make install \
    && apt-get purge -y openjdk-7-jdk="$JAVA_JDK_DEBIAN_VERSION" build-essential libssl-dev libapr1-dev \
    && apt-get -y autoremove \
    && rm -rf /tmp/tomcat-native*

ENV LD_LIBRARY_PATH /usr/local/apr/lib:$LD_LIBRARY_PATH
```

**Comment by md5** [comment]
@ahromis Yep. It might be nice to figure out how to make it install into `/usr/local/lib` so that the `LD_LIBRARY_PATH` change is not necessary. Either that or drop an `apr.conf` file into `/etc/ld.so.conf.d`.

**Comment by md5** [comment]
Also, I don't think `build-essential` should be necessary. That package is intended for building Debian packages and installs quite a bit more than I expect is needed to build `libtcnative`.

**Comment by andyshinn** [comment]
@ahromis I'm happy to test if you can post a full `Dockerfile`. This is something I'm looking to get fixed and have an application for it.

**Comment by ahromis** [comment]
Thanks @andyshinn I appreciate the help.  I'm curious to hear how it works out for you.  Can you try this out when you get a chance?  I only committed for Tomcat 7 JRE 7 as that's what I've been messing with today.

https://github.com/ahromis/tomcat/blob/7_add_tomcat_native_libraries/7-jre7/Dockerfile

**Comment by ahromis** [comment]
@md5 I made the changes you suggested.  `/usr/local/lib` isn't in the `java.library.path` so I used `/usr/lib/jni` instead.  Also `gcc` + `make` does the trick just fine and installs a lot cleaner, I appreciate the suggestion!

**Comment by mccoder** [comment]
#23 

Added pull request.
I compile tomcat-native and use latest version

Because libtcnative-1 old.

**Comment by rbjorklin** [comment]
Am I correct in believing that this would solve the following message?

    01-Mar-2016 14:18:41.036 INFO [main] org.apache.catalina.core.AprLifecycleListener.lifecycleEvent The APR based Apache Tomcat Native library which allows optimal performance in production environments was not found on the java.library.path: /usr/java/packages/lib/amd64:/usr/lib/x86_64-linux-gnu/jni:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/usr/lib/jni:/lib:/usr/lib

**Comment by mccoder** [comment]
@rbjorklin Yes.

The log is a message:
`Mar 01, 2016 4:08:01 PM org.apache.catalina.core.AprLifecycleListener lifecycleEvent
INFO: Loaded APR based Apache Tomcat Native library 1.1.34 using APR version 1.5.1.
Mar 01, 2016 4:08:01 PM org.apache.catalina.core.AprLifecycleListener lifecycleEvent
INFO: APR capabilities: IPv6 [true], sendfile [true], accept filters [false], random [true].
Mar 01, 2016 4:08:01 PM org.apache.catalina.core.AprLifecycleListener initializeSSL
INFO: OpenSSL successfully initialized (OpenSSL 1.0.1k 8 Jan 2015)
`

**Comment by jchambers** [comment]
It looks like this is resolved by #28; should this issue be closed?

**Comment by yosifkit** [comment]
Thanks, looks like we missed a "fixes".
